timelock/program/src/state/governance_voting_record.rs

```diff
@@ -0,0 +1,106 @@
+use super::UNINITIALIZED_VERSION;
+use arrayref::{array_mut_ref, array_ref, array_refs, mut_array_refs};
+use solana_program::{
+    program_error::ProgramError,
+    program_pack::{IsInitialized, Pack, Sealed},
+    pubkey::Pubkey,
+};
+
+/// STRUCT VERSION
+pub const GOVERNANCE_VOTING_RECORD_VERSION: u8 = 1;
+/// Governance Voting Record
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct GovernanceVotingRecord {
+    /// proposal
+    pub proposal: Pubkey,
+    /// owner
+    pub owner: Pubkey,
+    ///version
+    pub version: u8,
+    /// How many votes were unspent
+    pub undecided_count: u64,
+    /// How many votes were spent yes
+    pub yes_count: u64,
+    /// How many votes were spent no
+    pub no_count: u64,
+}
+
+impl Sealed for GovernanceVotingRecord {}
+impl IsInitialized for GovernanceVotingRecord {
+    fn is_initialized(&self) -> bool {
+        self.version != UNINITIALIZED_VERSION
+    }
+}
+
+/// Len of governance voting record
+pub const GOVERNANCE_VOTING_RECORD_LEN: usize = 32 + 32 + 1 + 8 + 8 + 8 + 100;
+impl Pack for GovernanceVotingRecord {
+    const LEN: usize = 32 + 32 + 1 + 8 + 8 + 8 + 100;
+    /// Unpacks a byte buffer into a [TimelockProgram](struct.TimelockProgram.html).
+    fn unpack_from_slice(input: &[u8]) -> Result<Self, ProgramError> {
```

Comment:
Definitely want to use something like Borsh so we can get rid of all this packing/unpacking code

Reply:
Oh yes, now that I know of Borsh, this is on my list of TODOs. Fear not, it's coming ;)

Reply:
Actually I've decided not to do borsh here. The front end JS lib really isn't working well, as it can't deserialize or serialize our Pubkeys and so it really isn't usable without a lot of hacking, and moreover our RPC calls where we use memcmp really doesn't work well when you never know where the actual bit offset will be, which is a constant worry with borsh-based structs. 

We personally lost a day or two just hacking with it to get it running for our NFT front end. Given it's not ready for primetime, I'd prefer to keep using Pack with this program and front end until that library accepts the pull request Jon Cinque made for it, and then we can circle back and discuss the use of borsh here.

Reply:
Ah crap, I meant to delete the other comment above to address this one below, and then lost my original comment. Oh well, rewriting! I don't think we should use borsh here. I recently used it in another project and the experience was less than desirable. While the backend experience is sublime, the front end experience is fairly immature. We ran into three major issues that cost considerable time to hack around:

1. RPC calls that use memcmp rely on bit offsets remaining consistent - you lose this with variable length structs
2. Figuring out what type a struct is from getProgramAccounts is impossible with borsh whereas it was quite easy when you could just compare a fixed length so you have to do try/catch unless you add some sort of identifier key
3. Borsh JS does not support deserialization or serialization of Pubkeys

All in all, the addition of borsh to the development process of the Metadata program cost us about 2 days of dev time. I don't want it to cost us further time here. Pack is verbose but in it's inelegance it is also perfectly defined and easy to work with. Moreover, it's already built. I'm more than happy to revisit the borsh discussion later, once the borsh-js team merges @joncinque 's PR (which has been sitting idle for 3 weeks now) that adds PublicKey support, but this still leaves the other 2 vulnerabilities in the design open. I personally think space is cheap while dev time is not - I'd rather have fixed length structs and more code than less code and variable length structs. I also dislike handing over my serde logic to an opaque library, because the amount of hours I spent debugging the descriptive Borsh::IOError was fairly large.

Reply:
Your pain points sound valid, and I hear you on it slowing down your development. A primary goal of SPL is to provide a set of polished reference programs that can be leveraged by the Solana community. We know that manual bit packing sucks in general and is error prone, and the wider community rightfully expects something more ergonomic. Throwing up our hands and feeding the ecosystem to the wolves isn't the right approach.

Some ideas to help us walk back from the ledge:
1. Fix Borsh if it can be.  If upstream isn't responsive, a short-lived fork that we maintain until upstream comes back is not ideal but not the worst
2. https://github.com/project-serum/anchor seems very promising and it's maintainer is 100% available to help and support. I would not be sad about dropping Borsh entirely for Anchor and then spending the time necessary to make Anchor work perfectly for our SPL use cases. It's ok if this causes the development of governance or other SPL candidates to be slightly slower.
3. Propose RPC API improvements (we own that API fully and can/will make improvements as demand requires)
 
Our pain is the ecosystem's gain and if we don't endure it nobody else will.

Reply:
Thanks @mvines, we will review each of the points and comeback with the proposal how to move this forward.